diff -urN lftp-3.7.4.org/m4/terminfo.m4 lftp-3.7.4/m4/terminfo.m4
--- lftp-3.7.4.org/m4/terminfo.m4	2004-02-12 14:37:26.000000000 +0100
+++ lftp-3.7.4/m4/terminfo.m4	2008-08-17 22:25:38.149166773 +0200
@@ -3,7 +3,7 @@
    AC_CHECK_HEADERS(curses.h term.h ncurses/curses.h ncurses/term.h)
 
    # Get a library with terminal caps if needed; prefer one with tigetstr.
-   AC_SEARCH_LIBS(tigetstr, [curses ncurses], [
+   AC_SEARCH_LIBS(tigetstr, [tinfo curses ncurses], [
 	 AC_DEFINE(HAVE_TIGETSTR, 1, [Define to 1 if you have the `tigetstr' function.])
 	 AC_CHECK_FUNCS(tgetstr)
       ], [AC_SEARCH_LIBS(tgetstr, termcap,
