--- lftp-4.0.0/configure.ac~	2009-09-12 19:39:22.434759846 +0200
+++ lftp-4.0.0/configure.ac	2009-09-12 19:39:56.352256686 +0200
@@ -254,9 +254,9 @@
 
 AC_ARG_WITH(gnutls, AS_HELP_STRING([--without-gnutls], [don't use GNUTLS library]),
       [with_gnutls=$withval], [with_gnutls=yes])
-AC_ARG_WITH(openssl,
-AS_HELP_STRING([--with-openssl@<:@=/path@:>@], [use OpenSSL @<:@at /path@:>@])
-AS_HELP_STRING([--without-openssl], [don't use OpenSSL (default)]),
+AC_ARG_WITH(openssl,[
+AS_HELP_STRING([--with-openssl@<:@=/path@:>@], [use OpenSSL @<:@at /path@:>@])
+AS_HELP_STRING([--without-openssl], [don't use OpenSSL (default)])],
       [with_openssl=$withval], [with_openssl=no])
 case "$with_openssl" in
 	yes)		with_gnutls=no;;
--- a/m4/ax_lib_expat.m4~	2016-02-20 14:57:52.000000000 +0100
+++ b/m4/ax_lib_expat.m4	2016-04-08 23:01:06.084309949 +0200
@@ -109,7 +109,6 @@ AC_DEFUN([AX_LIB_EXPAT],
 
     if test -n "$expat_prefix"; then
         expat_include_dir="$expat_prefix/include"
-        expat_ld_flags="-L$expat_prefix/lib"
         expat_lib_flags="-lexpat"
         run_expat_test="yes"
     elif test "$expat_requested" = "yes"; then
