diff -Nru lftp-2.3.9/src/Makefile.am lftp-2.3.9.new/src/Makefile.am
--- lftp-2.3.9/src/Makefile.am	Thu Mar 22 10:54:40 2001
+++ lftp-2.3.9.new/src/Makefile.am	Fri Apr 13 01:28:58 2001
@@ -62,10 +62,6 @@
 
 INCLUDES = -I$(top_srcdir)/include $(SSL_CPPFLAGS)
 
-# These files can be linked to ../include by configure, but not always.
-OMIT_DEPENDENCIES = readline.h keymaps.h chardefs.h tilde.h \
-	history.h libintl.hregex.h glob.h fnmatch.h poll.h rlstdc.h
-
 lftp.o commands.o module.o: confpaths.h
 
 
