--- lftp-2.6.9/configure.ac.orig	2003-11-20 16:21:37.000000000 +0100
+++ lftp-2.6.9/configure.ac	2003-11-20 23:54:06.499391944 +0100
@@ -282,7 +282,7 @@
 fi
 
 # See if the system has hstrerror, and replace it if not
-AC_CHECK_FUNC(hstrerror, , AC_LIBOBJ(hstrerror))
+AC_CHECK_FUNC(hstrerror, , [AC_LIBOBJ(hstrerror)])
 if test "$ac_cv_func_hstrerror" = yes; then
   IU_CHECK_DECL(hstrerror, [#include <netdb.h>])
 else
