--- lftp-2.5.0a/src/Makefile.am.orig	Fri Jan  4 12:33:32 2002
+++ lftp-2.5.0a/src/Makefile.am	Sun Apr 14 12:35:21 2002
@@ -12,7 +12,7 @@
 lftp_SOURCES = lftp.cc complete.h complete.cc lftp_rl.c lftp_rl.h
 #ftpget_SOURCES = ftpget.cc
 
-noinst_LIBRARIES = libtasks.a libjobs.a
+noinst_LTLIBRARIES = libtasks.la libjobs.la
 
 EXTRA_LTLIBRARIES = libnetwork.la proto-ftp.la proto-http.la proto-file.la cmd-mirror.la cmd-sleep.la proto-fish.la
 pkgverlib_LTLIBRARIES = $(MODULES_LA)
@@ -39,7 +39,7 @@
 proto_http_la_LIBADD = -L$(DESTDIR)$(pkgverlibdir) libnetwork.la
 proto_fish_la_LIBADD = -L$(DESTDIR)$(pkgverlibdir) libnetwork.la
 
-libtasks_a_SOURCES = PollVec.cc PollVec.h SMTask.cc SMTask.h ProcWait.cc\
+libtasks_la_SOURCES = PollVec.cc PollVec.h SMTask.cc SMTask.h ProcWait.cc\
  ProcWait.h ftpclass.cc ftpclass.h GetPass.cc GetPass.h\
  CharReader.cc CharReader.h LsCache.cc LsCache.h Resolver.cc Resolver.h\
  FileAccess.h FileAccess.cc ResMgr.h ResMgr.cc LocalAccess.h LocalAccess.cc\
@@ -58,7 +58,7 @@
  StringPool.cc StringPool.h DirColors.cc DirColors.h IdNameCache.cc\
  IdNameCache.h PatternSet.cc PatternSet.h LocalDir.cc LocalDir.h
 
-libjobs_a_SOURCES = Job.cc Job.h CmdExec.cc CmdExec.h\
+libjobs_la_SOURCES = Job.cc Job.h CmdExec.cc CmdExec.h\
  commands.cc mgetJob.h mgetJob.cc SysCmdJob.cc SysCmdJob.h rmJob.cc rmJob.h\
  parsecmd.cc MirrorJob.cc MirrorJob.h\
  mvJob.cc mvJob.h alias.cc alias.h\
@@ -89,7 +89,11 @@
 #OMIT_DEPENDENCIES = readline.h keymaps.h chardefs.h tilde.h rltypedefs.h\
 #	history.h libintl.hregex.h glob.h fnmatch.h poll.h rlstdc.h
 
-lftp.o commands.o module.o: confpaths.h
+lftp.lo commands.lo module.lo: confpaths.h
+libtasks.a: libtasks.la
+	ln -sf .libs/libtasks.a .
+libjobs.a: libjobs.la
+	ln -sf .libs/libjobs.a .
 
 # libtool does not strip modules, do it here.
 install-lib-local:
