--- ./src/complete.cc.orig	Mon May 14 20:59:10 2001
+++ ./src/complete.cc	Mon May 14 22:13:26 2001
@@ -38,8 +38,14 @@
 #include "lftp_rl.h"
 #include "url.h"
 
+#include <ctype.h>
+
 CDECL_BEGIN
+#undef __P
+#define __P(args) args
 #include "readline/readline.h"
+#undef __P
+#define __P(args) args __THROW
 CDECL_END
 
 static char *bash_dequote_filename (char *text, int quote_char);
@@ -577,7 +583,7 @@
    textq=bash_dequote_filename(textq, quoted);
    len=strlen(textq);
 
-   char **matches=lftp_rl_completion_matches(textq,generator);
+   char **matches=rl_completion_matches(textq,generator);
 
    glob_res=0;
    if(rg)
@@ -952,7 +958,6 @@
   return (0);
 }
 
-extern "C" int (*rl_last_func)(int,int);
 static int lftp_complete_remote(int count,int key)
 {
 
