--- ./src/Makefile.am.orig	Mon May 14 20:47:34 2001
+++ ./src/Makefile.am	Mon May 14 20:48:00 2001
@@ -82,8 +82,8 @@
 INCLUDES = -I$(top_srcdir)/include $(SSL_CPPFLAGS)
 
 # These files can be linked to ../include by configure, but not always.
-OMIT_DEPENDENCIES = readline.h keymaps.h chardefs.h tilde.h rltypedefs.h\
-	history.h libintl.hregex.h glob.h fnmatch.h poll.h rlstdc.h
+#OMIT_DEPENDENCIES = readline.h keymaps.h chardefs.h tilde.h rltypedefs.h\
+#	history.h libintl.hregex.h glob.h fnmatch.h poll.h rlstdc.h
 
 lftp.o commands.o module.o: confpaths.h
 
--- ./src/Makefile.in.orig	Mon May 14 20:49:08 2001
+++ ./src/Makefile.in	Mon May 14 20:49:27 2001
@@ -173,7 +173,7 @@
 INCLUDES = -I$(top_srcdir)/include $(SSL_CPPFLAGS)
 
 # These files can be linked to ../include by configure, but not always.
-OMIT_DEPENDENCIES = readline.h keymaps.h chardefs.h tilde.h rltypedefs.h 	history.h libintl.hregex.h glob.h fnmatch.h poll.h rlstdc.h
+#OMIT_DEPENDENCIES = readline.h keymaps.h chardefs.h tilde.h rltypedefs.h 	history.h libintl.hregex.h glob.h fnmatch.h poll.h rlstdc.h
 
 mkinstalldirs = $(SHELL) $(top_srcdir)/mkinstalldirs
 CONFIG_HEADER = ../include/config.h
